### PR TITLE
Topic/drawable extrusion

### PR DESCRIPTION
**Pull request by TimSylvester**
Mostly working.  The legacy path drops out sometimes, and it seems to follow tile boundaries so is probably stencil-related, but that might just be a problem with the split view.

Likely related to #1286.

### CODE REVIEW

**Review by sjg-wdw — APPROVED** [review]

**Review by stefankarschti — APPROVED** [review]

**Comment by mwilsnd** [comment]
If I rotate the map, I can see the lighting/shading doesn't update correctly, ex:
image

**Comment by TimSylvester** [comment]
light parameters are updated every frame now

File: src/mbgl/renderer/layers/render_fill_extrusion_layer.cpp [hunk 1]
```diff
@@ -241,4 +276,227 @@ bool RenderFillExtrusionLayer::queryIntersectsFeature(const GeometryCoordinates&
...
+
+        if (doDepthPass && !depthBuilder) {
+            if (auto builder = context.createDrawableBuilder(layerPrefix + "depth")) {
+                builder->setShader(shader);
+                builder->setIs3D(true);
+                builder->setEnableStencil(false);
+                builder->setEnableColor(false);
+                builder->setRenderPass(drawPass);
+                builder->setCullFaceMode(gfx::CullFaceMode::backCCW());
+                depthBuilder = std::move(builder);
```
Comment (mwilsnd) [inline src/mbgl/renderer/layers/render_fill_extrusion_layer.cpp:420]: I see the reason why we have two builders but the extra copies we're doing here just to have different state isn't great. Can we have a single builder and get shared vertex buffers working? ie Two or more drawables sharing the same VAO but configured with different state.

File: src/mbgl/gfx/fill_extrusion_drawable_data.hpp [hunk 2]
```diff
@@ -0,0 +1,20 @@
+#pragma once
```
Comment (mwilsnd) [inline src/mbgl/gfx/fill_extrusion_drawable_data.hpp:1]: Is this drawable data being used or do you have plans for it to be used?

File: src/mbgl/renderer/layers/render_fill_extrusion_layer.cpp (hunk 1 above)
Comment (sjg-wdw) [reply src/mbgl/renderer/layers/render_fill_extrusion_layer.cpp:420]: If we are going to do that, can we do it at a separate task?  Be happy to schedule it very soon.

Comment (mwilsnd) [reply src/mbgl/renderer/layers/render_fill_extrusion_layer.cpp:420]: I'll make an issue for it.

Comment (mwilsnd) [reply src/mbgl/renderer/layers/render_fill_extrusion_layer.cpp:420]: Tracking this in #1300, OK to leave this for now.

File: src/mbgl/gfx/fill_extrusion_drawable_data.hpp (hunk 2 above)
Comment (TimSylvester) [reply src/mbgl/gfx/fill_extrusion_drawable_data.hpp:1]: No, it turned out to not be needed, and I forgot about it.

**Review by mwilsnd — APPROVED** [review]

File: src/mbgl/gfx/drawable_builder.cpp [hunk 3]
```diff
@@ -34,12 +34,15 @@ void DrawableBuilder::flush() {
...
         draw->setLineWidth(static_cast<int32_t>(lineWidth));
-        draw->setNeedsStencil(needsStencil);
+        draw->setEnableColor(enableColor);
+        draw->setEnableStencil(enableStencil);
         draw->setRenderPass(renderPass);
         draw->setDrawPriority(drawPriority);
         draw->setSubLayerIndex(subLayerIndex);
         draw->setDepthType(depthType);
+        draw->setIs3D(is3D);
         draw->setCullFaceMode(impl->cullFaceMode);
```
Comment (alexcristici) [inline src/mbgl/gfx/drawable_builder.cpp:46]: If you merge this after the heatmap layer, you will have to add `setColorMode`.

**Review by alexcristici — APPROVED** [review]

**Pull request merged**